File: output/llama3.1-8b/avatar/Python/Java/codeforces_216_B.java
import java.util.*;

public class codeforces_216_B {
    public static int[] readInts() {
        String[] input = scanner.nextLine().split(" ");
        int[] result = new int[input.length];
        for (int i = 0; i < input.length; i++) {
            result[i] = Integer.parseInt(input[i]);
        }
        return result;
    }

    public static void dfs(Map<Integer, Set<Integer>> g, Set<Integer> seen, int i) {
        if (seen.contains(i)) {
            return;
        }
        seen.add(i);
        int nodes = 1;
        int edges = g.get(i).size();
        for (int j : g.get(i)) {
            int[] x = dfs(g, seen, j);
            nodes += x[0];
            edges += x[1];
        }
    }

    public static boolean solve() {
        try {
            int[] line0 = readInts();
            int n = line0[0], m = line0[1];
            Map<Integer, Set<Integer>> g = new HashMap<>();
            Set<Integer> seen = new HashSet<>();
            for (int i = 1; i <= n; i++) {
                g.put(i, new HashSet<>());
            }
            for (int i = 0; i < m; i++) {
                int[] ab = readInts();
                g.get(ab[0]).add(ab[1]);
                g.get(ab[1]).add(ab[0]);
            }
            int ans = 0;
            for (int i = 1; i <= n; i++) {
                if (!seen.contains(i)) {
                    int[] x = dfs(g, seen, i);
                    if (x[0] > 1 && x[0] % 2 == 1 && 2 * x[0] == x[1]) {
                        ans++;
                    }
                }
            }
            if ((n - ans) % 2 == 1) {
                ans++;
            }
            System.out.println(ans);
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    public static void main(String[] args) {
        while (solve()) ;
    }
}

File: output/llama3.1-8b/avatar/Java/Python/codeforces_216_B.py
from collections import defaultdict, deque

class Graph:
    def __init__(self, n):
        self.n = n
        self.graph = [[] for _ in range(n + 1)]
        self.seen = [False] * (n + 1)
        self.count = 0

    def dfs(self, child, par):
        if self.seen[child]:
            return 2
        self.seen[child] = True
        for i in self.graph[child]:
            if i != par:
                self.count += 1
                if self.dfs(i, child) == 2:
                    return 2
        return 0

    def solve(self):
        m, n = map(int, input().split())
        for _ in range(m + 1):
            self.graph.append([])
        for _ in range(n):
            x, y = map(int, input().split())
            self.graph[x].append(y)
            self.graph[y].append(x)

        toremove = 0
        for i in range(1, m + 1):
            if not self.seen[i]:
                self.count = 0
                if self.dfs(i, 0) == 2:
                    if self.count % 2 == 1:
                        toremove += 1

        if (m - toremove) % 2 == 1:
            toremove += 1
        print(toremove)


if __name__ == "__main__":
    g = Graph(int(input()))
    g.solve()